### PR TITLE
balena device: Add ability to move multiple devices in one command

### PR DESCRIPTION
**Pull request by nwneisen**
Add the ability to move multiple devices in one command

Partially Resolves: #323 
Change-type: minor

### CODE REVIEW

File: lib/actions-oclif/device/move.ts [hunk 1]
```diff
@@ -82,47 +83,51 @@ export default class DeviceMoveCmd extends Command {
...
-			const compatibleDeviceTypes = deviceTypes.filter(
-				(dt) =>
-					balena.models.os.isArchitectureCompatibleWith(
-						deviceDeviceType.arch,
-						dt.arch,
-					) &&
-					!!dt.isDependent === !!deviceDeviceType.isDependent &&
-					dt.state !== 'DISCONTINUED',
+		for (const uuid of params.uuid.split(',')) {
+			const device: ExtendedDevice = await balena.models.device.get(
```
Comment (pdcastro) [inline lib/actions-oclif/device/move.ts:88]: If I'm reading this right (I may not be!), a network request for `await balena.models.device.get(...)` is made for each `uuid`, but the resulting `device` object is only actually possibly used for the first device, in order to determine the destination application (and even then, only in case `options.application` was not provided). Because ultimately, to actually move the device, all that is needed seems to be uuid string:
```
    await balena.models.device.move(uuid, tryAsInteger(application));
```

So how about the following structure:
* Extract the first `uuid` (first device), before any `for` loop, something like:  
`const uuid = params.uuid.split(',')[0];`
* Then determine the destination application. Only if needed (if `options.application` was not provided), call `await balena.models.device.get()` for that first uuid only.
* Then, knowing the destination application, have a small for loop moving the devices:
```
	for (const uuid of params.uuid.split(',')) {
		await balena.models.device.move(uuid, tryAsInteger(application));
		console.info(`${uuid} was moved to ${application}`);
	}
```

File: lib/actions-oclif/device/move.ts [hunk 2]
```diff
@@ -82,47 +83,51 @@ export default class DeviceMoveCmd extends Command {
...
-					// @ts-ignore using the extended device object prop
-					device.application_name !== app.app_name,
-			);
+			const belongsToApplication = device.belongs_to__application as Application[];
+			device.application_name = belongsToApplication?.[0]
+				? belongsToApplication[0].app_name
+				: 'N/a';
+
+			// Get destination application
+			if (!application) {
```
Comment (pdcastro) [inline lib/actions-oclif/device/move.ts:99]: My other comment suggests a different overall structure. If that structure was not feasible, then I'm thinking it would easier to read the code if this `if (!application) {...}` block was move outside the `for` loop, perhaps in a function.

File: lib/actions-oclif/device/move.ts [hunk 3]
```diff
@@ -40,28 +40,29 @@ interface ArgsDef {
...
 		'$ balena device move 7cf02a6',
+		'$ balena device move 7cf02a6,dc39e52',
 		'$ balena device move 7cf02a6 --application MyNewApp',
 	];
 
 	public static args: Array<IArg<any>> = [
 		{
 			name: 'uuid',
-			description: 'the uuid of the device to move',
+			description: 'the uuid(s) of the device to move',
```
Comment (pdcastro) [inline lib/actions-oclif/device/move.ts:59]: Suggestion:
```
description: 'comma-separated list (no blank spaces) of device UUIDs to be moved',
```

File: lib/actions-oclif/device/move.ts (hunk 1 above)
Comment (nwneisen) [reply lib/actions-oclif/device/move.ts:88]: The one problem I see with this method is that `await balena.models.device.get(...)` is validating the uuids exist. If one of the uuids is invalid then the operation will stop before anything is moved. Without this the operation could throw an error and stop with half of the devices moved and the second half left as is. Perhaps the solution for this would be report the uuid that failed to the user and continue trying to move the remaining uuids.

Comment (pdcastro) [reply lib/actions-oclif/device/move.ts:88]: That's a good point and feature: validating that all UUIDs exist before attempting to move any of them. 👍 
In this case, how about:
* Query all UUIDs, perhaps even in parallel with `Promise.all`:  
`const devices = await Promise.all(params.uuid.split(',').map((uuid) => balena.models.device.get(...)));`
* Then determine the destination application, using `devices[0]` as needed (if `options.application` was not provided).
* Then, knowing the destination application, have a small for loop moving the devices:
```
	for (const uuid of params.uuid.split(',')) {
		await balena.models.device.move(uuid, tryAsInteger(application));
		console.info(`${uuid} was moved to ${application}`);
	}
```

File: lib/actions-oclif/device/move.ts [hunk 4]
```diff
@@ -82,47 +85,67 @@ export default class DeviceMoveCmd extends Command {
...
-			const compatibleDeviceTypes = deviceTypes.filter(
-				(dt) =>
-					balena.models.os.isArchitectureCompatibleWith(
-						deviceDeviceType.arch,
-						dt.arch,
-					) &&
-					!!dt.isDependent === !!deviceDeviceType.isDependent &&
-					dt.state !== 'DISCONTINUED',
+			const compatibleDeviceTypes = deviceTypes.filter((dt) =>
+				deviceDeviceTypes.every(
```
Comment (pdcastro) [inline lib/actions-oclif/device/move.ts:116]: I am contributing a commit: please review it, test the command and feel free to make any further changes (and let me know when/if you're happy. I was doing some testing myself and, before my commit, I realised that some corner cases existed if the device list mixed devices of incompatible architectures. Example, before my commit:
```
$ ./bin/balena device move b1878c6,dcbff32
? Select an application
❯ test-19282 (qemux86-64)
  test-nuc (intel-nuc)

$ ./bin/balena device move dcbff32,b1878c6
? Select an application
  tests (raspberrypi3)
  test-skx2 (skx2)
  test-tx2 (jetson-tx2)
❯ bbb-test (beaglebone-green-wifi)
  essentials (raspberrypi3)
  iwantlegacy (raspberrypi3)
  raspberrypi2 (raspberry-pi2)
```
Where b1878c6 is an amd64 device, dcbff32 is an ARM device. Note how the list of applications changes depending on the order of the devices in the comma-separated list in the command line. By the way, I realise that it was my own suggestion to use `devices[0]`, then again, the original code before my suggestion was also subject to this issue. :-)
After my commit:
```
$ ./bin/balena device move dcbff32,b1878c6
No suitable applications found for selection
Do all devices have a compatible architecture?

$ ./bin/balena device move dcbff32,b1878c6 -a test-rpi
dcbff32 was moved to test-rpi
BalenaInvalidDeviceType: Invalid device type: Incompatible application: test-rpi
```
The last case above raises `BalenaInvalidDeviceType`, which I had added to `errors.ts` as an "expected error". I think it is acceptable.

File: lib/actions-oclif/device/move.ts [hunk 5]
```diff
@@ -144,8 +144,12 @@ export default class DeviceMoveCmd extends Command {
 		}
 
 		for (const uuid of params.uuid.split(',')) {
-			await balena.models.device.move(uuid, tryAsInteger(application));
-			console.info(`${uuid} was moved to ${application}`);
+			try {
```
Comment (nwneisen) [inline lib/actions-oclif/device/move.ts:147]: I found an additional error case with both of our code changes. When passing the application with `-a`, valid devices before an invalid arch uuid will be moved and devices after will not.This change will move all devices with a valid arch and report the uuid of invalid arch devices. 

This is different behavior than we previously discussed in which we validate devices before moving them. I think moving valid devices and reporting invalid or not moving any when an invalid is found could both be appropriate behavior as long as it's clear to the user. The down side of the first method is this is calling the move command and expecting an error during normal behavior.

I'd like to change method to use the following process:

1. Acquire and validate(not currently done for `-a`) the application.
2. Acquire valid device arch's for the application.
3. Verify all device arch's are valid for the application. Report those that aren't.
4. Move devices if all device archs are valid.

I looked around but wasn't able to find a method to validate an application exists by string or get a list of valid archs for the application.

File: lib/actions-oclif/device/move.ts [hunk 6]
```diff
@@ -144,8 +144,12 @@ export default class DeviceMoveCmd extends Command {
...
+				await balena.models.device.move(uuid, tryAsInteger(application));
+				console.info(`${uuid} was moved to ${application}`);
+			} catch (err) {
+				console.info(`${err.message}, uuid: ${uuid}`);
```
Comment (pdcastro) [inline lib/actions-oclif/device/move.ts:151]: Nice! 👍   Additional easy improvement: add `process.exitCode = 1;` here, so that the CLI process exit code reflects the fact that something went wrong and some devices were not moved (in case the CLI is being scripted).

> I'd like to change method to use the following process: [...] Move devices if all device archs are valid.

You're welcome to, but my opinion/suggestion is that it is good enough as it is, and it could always be improved further in the future (in the way you suggest). I think it is already a great feature delivery as it is, and additional PRs to deliver the same to other commands (e.g. `device rm`) (or PRs for other issues) would offer a greater return on investment at this point! :-)

Comment (nwneisen) [reply lib/actions-oclif/device/move.ts:151]: `process.exitCode = 1` added. I've also got a branch started to add the same functionality to `device rm`.

Comment (pdcastro) [reply lib/actions-oclif/device/move.ts:151]: Great! Would you mind squashing the 4 commits in a single one? Probably using the description of your first commit. Generally we follow the principle of "meaningful commits", and the commits in this PRs at the moment are effectively amending each other.

Comment (nwneisen) [reply lib/actions-oclif/device/move.ts:151]: I debated about doing so but waited to see what was typically done for the project.

**Comment by pdcastro** [comment]
@balena-ci retest

File: lib/actions-oclif/device/move.ts (hunk 6 above)
Comment (pdcastro) [reply lib/actions-oclif/device/move.ts:151]: Oh, sorry, just noticed one more thing: the `Change-type` for the commit should be `minor` rather than `patch`, as it is delivering a new feature (and one that is visible to users!).

**Comment by pdcastro** [comment]
@balena-ci retest

**Comment by pdcastro** [comment]
> _It looks like balena CI is ignoring the rebase command, although it understood the retest command. Not sure if it is specific to this PR: I'll ask the CI team to have a look._

Re comment above (from PR #1935), the CI team thinks that is because the source branch is in a forked repo -- nothing wrong with it, it would be a balena CI bug. An issue was created to get this fixed. I have also noticed something else (another CI bug): when you push or force-push a new commit, tests are not re-triggered automatically. Chances are it is the same cause: forked repo. I'll report to the CI team as well. Until then, it looks like the reliable (?) way for you to get the tests re-triggered is to add the `(at)balena-ci retest` command in a PR comment (I had done it, and then again after the force push).

**Comment by nwneisen** [comment]
> the source branch is in a forked repo
> Chances are it is the same cause: forked repo

Out of curiosity what is the alternative method for making changes? Creating branches directly on the main repo? I had thought it was standard practice on most projects to create a fork and merge branches back into the main repo.

**Comment by pdcastro** [comment]
> Creating branches directly on the main repo?

Yes.

> I had thought it was standard practice on most projects to create a fork and merge branches back into the main repo.

I have also previously worked like that (in another company), but I don't really about what the "standard" (most popular?) practice is.

**Comment by pdcastro** [comment]
By the way, talking of odd things, I've also noticed that the automatically updated Changelog uses your first name only, whereas everybody else has name and surname listed there:
https://github.com/balena-io/balena-cli/blob/master/CHANGELOG.md
```
12.8.0 - 2020-07-15 device move: Add ability to move multiple devices in one command [Nick]
12.7.7 - 2020-07-14 Remove os download command dependency for device init [Scott Lowe]
```
Would this be some configuration in your GitHub account? Yet, your surname is clearly visible on the following page:
https://github.com/nwneisen

Would it be that balena CI doesn't have some permissions? No idea...

**Comment by nwneisen** [comment]
@pdcastro My local machine only had my first name set in the git credentials. I've updated it to be my full name and we will find out on my next commit if that was the reason.
